### PR TITLE
Fix function jsonMergePatch for partially const columns

### LINKED ISSUE #56384: jsonMergePatch in combination with JSONExtractArrayRaw throws error
version 23.10.1.1976  

I'm testing the new `jsonMergePatch` function and trying to add an element to an array does not seem to work when using JSON function `JSONExtractArrayRaw`.

Starting with `{"id":1,"foo":["bar"]}`.
Desired result: `{"id":1,"foo":["bar","baz"]}`.

```sql
-- works  
select '{"id":1,"foo":["bar"]}' as a, jsonMergePatch(a,'{"foo":["bar","baz"]}') as b;
select '{"id":1,"foo":["bar"]}' as a, jsonMergePatch(a,toJSONString(map('foo',['bar','baz']))) as b;
select '{"id":1,"foo":["bar"]}' as a, jsonMergePatch(a,toJSONString(map('foo',arrayPushBack(['bar'],'baz')))) as b;
  
-- expected to work but doesn't: Error Code: 44. DB::Exception: Arguments of function jsonMergePatch must be strings. (ILLEGAL_COLUMN) (version 23.10.1.1976 (official build))
select '{"id":1,"foo":["bar"]}' as a, jsonMergePatch(a,toJSONString(map('foo',arrayPushBack(arrayMap(x->JSONExtractString(x),JSONExtractArrayRaw(a, 'foo')),'baz')))) as b;
  
-- this workaround works
with 
  '{"id":1,"foo":["bar"]}' as a,
  (select toJSONString(map('foo',arrayPushBack(arrayMap(x->JSONExtractString(x),JSONExtractArrayRaw(a, 'foo')),'baz')))) as newfoo
select jsonMergePatch(a,newfoo) as b
;
```

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix function jsonMergePatch for partially const columns. Closes https://github.com/ClickHouse/ClickHouse/issues/56384

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit c8d964865eb57afcb794ebd7428376365a8d85e5 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57379/c8d964865eb57afcb794ebd7428376365a8d85e5/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7041946913">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57379/c8d964865eb57afcb794ebd7428376365a8d85e5/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by evillique** [comment]
Tests:
- Stateless tests (release, s3 storage) [2/2]: https://github.com/ClickHouse/ClickHouse/issues/57444

File: src/Functions/jsonMergePatch.cpp [hunk 1]
```diff
@@ -102,8 +102,11 @@ namespace
                     throw Exception(ErrorCodes::BAD_ARGUMENTS, "Wrong JSON string to merge. Expected JSON object");
             };
 
-            const auto * first_string = typeid_cast<const ColumnString *>(arguments[0].column.get());
-            if (!first_string)
+            const auto * first_column_arg = arguments[0].column.get();
+            const auto * first_column_arg_string = typeid_cast<const ColumnString *>(first_column_arg);
+            const auto * first_column_arg_const = typeid_cast<const ColumnConst *>(first_column_arg);
```
Comment (Avogar) [inline src/Functions/jsonMergePatch.cpp:107]: Should we also check that nested column is String?

Comment (Avogar) [reply src/Functions/jsonMergePatch.cpp:107]: I would do smth like this:
```
bool is_first_const = isColumnConst(*arguments[0].column);
const auto * first_column_arg_string = is_first_const
            ? checkAndGetColumnConstData<ColumnString>(arguments[0].column.get())
            : checkAndGetColumn<ColumnString>(arguments[0].column.get());
```
And later pass always 0 index to `parse_json_document` if `is_first_const=true`

Comment (Avogar) [reply src/Functions/jsonMergePatch.cpp]: Well, now we pass not `ColumnString` to the lambda, but `IColumn`, so we call virtual method `getDataAt`, and devirtualization on each row can be expensive

Comment (Avogar) [reply src/Functions/jsonMergePatch.cpp]: I may be wrong, if you think current way is ok, then no problem

**Review by Avogar — APPROVED** [review]

**Comment by evillique** [comment]
Tests:
- Integration tests (release) [4/4]: https://github.com/ClickHouse/ClickHouse/issues/54502
- Stateless tests (release, s3 storage) [2/2]: https://github.com/ClickHouse/ClickHouse/issues/57444

**Pull request merged**